TourMate/TourMate/Backend/Authentication/Managers/FirebaseAuthenticationManager.swift

```diff
@@ -10,25 +10,45 @@ import Firebase
 import FirebaseAuth
 import GoogleSignIn
 
-// https://peterfriese.dev/posts/firebase-async-calls-swift/
-struct FirebaseAuthenticationManager: AuthenticationManager {
+class FirebaseAuthenticationManager: AuthenticationManager {
 
     private let userService: UserService
 
+    private var authStateListenerHandle: AuthStateDidChangeListenerHandle?
+    weak var firebaseAuthDelegate: FirebaseAuthenticationDelegate?
+
     init(userService: UserService) {
         self.userService = userService
     }
 
-    func checkIfUserIsLoggedIn() -> Bool {
-        Auth.auth().currentUser != nil
+    func getCurrentFirebaseUser() -> Firebase.User? {
+        Auth.auth().currentUser
+    }
+
+    func fetchLogInStateAndListen() {
+        self.authStateListenerHandle = Auth.auth().addStateDidChangeListener { _, user in
+            self.firebaseAuthDelegate?.update(isLoggedIn: user != nil)
+        }
+    }
+
+    func detachListener() {
+        if let handle = authStateListenerHandle {
+            Auth.auth().removeStateDidChangeListener(handle)
+        }
+        authStateListenerHandle = nil
     }
 
     func logIn() {
+        print("[FirebaseAuthenticationManager] logging in")
```

Comment:
I noticed that the logged in method accesses UI stuff, not sure if there's other ways to do it because this seems to violate a lot of SoC.

Reply:
Also it seems like its getting runtime warnings here

<img width="1841" alt="image" src="https://user-images.githubusercontent.com/65328119/162601905-a8d6b1b4-e37f-45b5-a4db-5c4e14621e73.png">

Reply:
I don't think this can be changed because the Google login is the one handling the view.